File: src/git/discovery.zig
//! Git rev-parse path discovery used to locate the Repository Store.
//!
//! `tk init` (and future `tk worktree` commands) need Git's common directory
//! and toplevel to place or find `<git-common-dir>/tk/ticket.db` (the
//! Repository Store, per ARCHITECTURE.md). This module owns the
//! `git rev-parse` invocation and returns a typed `Outcome` so callers can
//! render diagnostics without this module reaching for stderr.

const std = @import("std");
const proc = @import("../proc/runner.zig");
const messages = @import("../messages.zig");

const Allocator = std.mem.Allocator;

/// Paths reported by Git for the current repository.
///
/// Both fields are heap-allocated by `discoverPaths` from the caller-supplied
/// allocator. Callers extract this from `Outcome.ok` and free it via
/// `DiscoveredPaths.deinit`.
pub const DiscoveredPaths = struct {
    /// Shared Git common directory; used as the parent of the Repository Store.
    git_common_dir: []u8,
    /// Repository working-tree root; used only for display_prefix derivation.
    toplevel: []u8,

    pub fn deinit(self: *DiscoveredPaths, gpa: Allocator) void {
        gpa.free(self.git_common_dir);
        gpa.free(self.toplevel);
    }
};

/// Outcome of `discoverPaths`. Each variant maps to a stable, caller-rendered
/// stderr message; this module never writes to stderr itself.
///
/// Ownership: each switch arm frees its own payload. `.ok` carries a
/// `DiscoveredPaths` whose two slices live on the caller's allocator and
/// must be released via `DiscoveredPaths.deinit`. `.git_rejected` may carry
/// an owned trimmed-stderr slice (also on the caller's allocator) which the
/// caller must `gpa.free` after rendering. The remaining variants carry no
/// payload. There is no `Outcome.deinit`; per-arm freeing keeps test and
/// production call sites identical.
pub const Outcome = union(enum) {
    ok: DiscoveredPaths,
    /// `git` is not on PATH (proc.Runner returned ExecutableNotFound).
    git_missing,
    /// `git` could not be spawned for a reason other than missing binary.
    spawn_failed,
    /// `git rev-parse` exited non-zero. Payload is the trimmed stderr if Git
    /// produced any (caller renders it verbatim and frees the slice); null
    /// means Git failed silently and the caller should render the default
    /// not-in-repo message.
    git_rejected: ?[]u8,
    /// `git rev-parse` exited zero but the stdout did not contain both
    /// expected lines. The "missing toplevel" sub-case is unreachable in
    /// practice because `--show-toplevel` always pairs with
    /// `--git-common-dir` when the repo is a worktree; the variant covers
    /// the missing-common-dir branch and serves as a defensive fallback.
    git_output_unparseable,
};

/// Errors returned by `discoverPaths` itself. Runner errors are mapped to
/// `Outcome` variants; only `OutOfMemory` from the allocator escapes.
pub const Error = error{OutOfMemory};

/// Run `git rev-parse --path-format=absolute --git-common-dir --show-toplevel`
/// through the supplied runner and classify the result.
pub fn discoverPaths(gpa: Allocator, runner: proc.Runner, cwd: std.Io.Dir) Error!Outcome {
    var run_result = runner.run(gpa, .{
        .argv = &.{ "git", "rev-parse", "--path-format=absolute", "--git-common-dir", "--show-toplevel" },
        .cwd = cwd,
    }) catch |err| switch (err) {
        error.OutOfMemory => return error.OutOfMemory,
        error.ExecutableNotFound => return .git_missing,
        error.SpawnFailed => return .spawn_failed,
    };
    defer run_result.deinit(gpa);

    if (run_result.exit_code != 0) {
        const trimmed = std.mem.trim(u8, run_result.stderr, " \t\r\n");
        if (trimmed.len == 0) return .{ .git_rejected = null };
        const owned = try gpa.dupe(u8, trimmed);
        return .{ .git_rejected = owned };
    }

    var lines = std.mem.tokenizeScalar(u8, run_result.stdout, '\n');
    const common = lines.next() orelse return .git_output_unparseable;
    const toplevel = lines.next() orelse return .git_output_unparseable;

    const common_owned = try gpa.dupe(u8, std.mem.trim(u8, common, " \t\r\n"));
    errdefer gpa.free(common_owned);
    const toplevel_owned = try gpa.dupe(u8, std.mem.trim(u8, toplevel, " \t\r\n"));
    errdefer gpa.free(toplevel_owned);

    return .{ .ok = .{ .git_common_dir = common_owned, .toplevel = toplevel_owned } };
}

/// Render an Outcome failure to stderr with a `tk <command>: ` prefix and free
/// any heap-owned payload it carries. Callers branch on `.ok` themselves
/// before delegating here; this helper exists to keep the four discovery
/// failure arms identical across every command that opens a Repository Store.
///
/// `command` is the bare subcommand name (`"init"`, `"add"`), without the `tk`
/// or the trailing colon — those are formatted in.
pub fn renderFailure(
    stderr: *std.Io.Writer,
    gpa: Allocator,
    command: []const u8,
    outcome: Outcome,
) void {
    switch (outcome) {
        .ok => unreachable,
        .git_missing => stderr.print("tk {s}: {s}\n", .{ command, messages.init_git_missing }) catch {},
        .spawn_failed => stderr.print("tk {s}: {s}\n", .{ command, messages.init_git_spawn_failed }) catch {},
        .git_rejected => |maybe_msg| {
            if (maybe_msg) |msg| {
                defer gpa.free(msg);
                stderr.print("tk {s}: {s}\n", .{ command, msg }) catch {};
            } else {
                stderr.print("tk {s}: {s}\n", .{ command, messages.init_outside_git_default }) catch {};
            }
        },
        .git_output_unparseable => stderr.print("tk {s}: {s}\n", .{ command, messages.init_git_unparseable }) catch {},
    }
}

// ---- Tests ---------------------------------------------------------------

const fake_proc = @import("../proc/fake.zig");

test "discoverPaths: returns .ok with both paths on success" {
    const gpa = std.testing.allocator;
    var fake = fake_proc.FakeRunner.init(gpa);
    defer fake.deinit();
    try fake.expect(&.{ "git", "rev-parse" }, .{
        .exit_code = 0,
        .stdout = "/repo/.git\n/repo\n",
    });

    const outcome = try discoverPaths(gpa, fake.runner(), std.Io.Dir.cwd());
    if (outcome != .ok) return error.UnexpectedOutcome;
    var paths = outcome.ok;
    defer paths.deinit(gpa);
    try std.testing.expectEqualStrings("/repo/.git", paths.git_common_dir);
    try std.testing.expectEqualStrings("/repo", paths.toplevel);
}

test "discoverPaths: returns .git_rejected with trimmed stderr on exit 128" {
    const gpa = std.testing.allocator;
    var fake = fake_proc.FakeRunner.init(gpa);
    defer fake.deinit();
    try fake.expect(&.{ "git", "rev-parse" }, .{
        .exit_code = 128,
        .stderr = "fatal: not a git repository (or any of the parent directories): .git\n",
    });

    const outcome = try discoverPaths(gpa, fake.runner(), std.Io.Dir.cwd());
    switch (outcome) {
        .git_rejected => |maybe_msg| {
            const msg = maybe_msg orelse return error.ExpectedNonNullMessage;
            defer gpa.free(msg);
            // Pin the exact trimmed value: trailing newline stripped, no
            // other modifications. Drift in trim semantics fails here.
            try std.testing.expectEqualStrings(
                "fatal: not a git repository (or any of the parent directories): .git",
                msg,
            );
        },
        else => return error.UnexpectedOutcome,
    }
}

test "discoverPaths: returns .git_rejected with null payload when stderr is whitespace" {
    const gpa = std.testing.allocator;
    var fake = fake_proc.FakeRunner.init(gpa);
    defer fake.deinit();
    try fake.expect(&.{ "git", "rev-parse" }, .{ .exit_code = 128, .stderr = "  \r\n  " });

    const outcome = try discoverPaths(gpa, fake.runner(), std.Io.Dir.cwd());
    switch (outcome) {
        .git_rejected => |maybe_msg| try std.testing.expect(maybe_msg == null),
        else => return error.UnexpectedOutcome,
    }
}

test "discoverPaths: returns .git_output_unparseable when stdout has no lines" {
    const gpa = std.testing.allocator;
    var fake = fake_proc.FakeRunner.init(gpa);
    defer fake.deinit();
    try fake.expect(&.{ "git", "rev-parse" }, .{ .exit_code = 0, .stdout = "" });

    const outcome = try discoverPaths(gpa, fake.runner(), std.Io.Dir.cwd());
    try std.testing.expect(outcome == .git_output_unparseable);
}

test "discoverPaths: returns .git_output_unparseable when stdout has only one line" {
    const gpa = std.testing.allocator;
    var fake = fake_proc.FakeRunner.init(gpa);
    defer fake.deinit();
    try fake.expect(&.{ "git", "rev-parse" }, .{ .exit_code = 0, .stdout = "/repo/.git\n" });

    const outcome = try discoverPaths(gpa, fake.runner(), std.Io.Dir.cwd());
    try std.testing.expect(outcome == .git_output_unparseable);
}

test "discoverPaths: maps ExecutableNotFound to .git_missing" {
    const gpa = std.testing.allocator;
    var injector = fake_proc.ErrorInjectingRunner{ .err = error.ExecutableNotFound };

    const outcome = try discoverPaths(gpa, injector.runner(), std.Io.Dir.cwd());
    try std.testing.expect(outcome == .git_missing);
}

test "discoverPaths: maps SpawnFailed to .spawn_failed" {
    const gpa = std.testing.allocator;
    var injector = fake_proc.ErrorInjectingRunner{ .err = error.SpawnFailed };

    const outcome = try discoverPaths(gpa, injector.runner(), std.Io.Dir.cwd());
    try std.testing.expect(outcome == .spawn_failed);
}

test "discoverPaths: propagates OutOfMemory from runner" {
    var injector = fake_proc.ErrorInjectingRunner{ .err = error.OutOfMemory };

    try std.testing.expectError(
        error.OutOfMemory,
        discoverPaths(std.testing.allocator, injector.runner(), std.Io.Dir.cwd()),
    );
}

test "renderFailure: formats `tk <command>: <fragment>` for each failure arm" {
    const gpa = std.testing.allocator;
    var buf: std.Io.Writer.Allocating = .init(gpa);
    defer buf.deinit();

    renderFailure(&buf.writer, gpa, "init", .git_missing);
    try std.testing.expectEqualStrings("tk init: " ++ messages.init_git_missing ++ "\n", buf.written());

    buf.clearRetainingCapacity();
    renderFailure(&buf.writer, gpa, "add", .spawn_failed);
    try std.testing.expectEqualStrings("tk add: " ++ messages.init_git_spawn_failed ++ "\n", buf.written());

    buf.clearRetainingCapacity();
    renderFailure(&buf.writer, gpa, "init", .git_output_unparseable);
    try std.testing.expectEqualStrings("tk init: " ++ messages.init_git_unparseable ++ "\n", buf.written());

    buf.clearRetainingCapacity();
    renderFailure(&buf.writer, gpa, "init", .{ .git_rejected = null });
    try std.testing.expectEqualStrings("tk init: " ++ messages.init_outside_git_default ++ "\n", buf.written());
}

test "renderFailure: frees the git_rejected payload it was handed" {
    // The testing allocator detects leaks, so if renderFailure forgets to
    // free the owned message slice this test fails. The test also pins that
    // the payload is rendered verbatim with the command prefix.
    const gpa = std.testing.allocator;
    var buf: std.Io.Writer.Allocating = .init(gpa);
    defer buf.deinit();

    const owned = try gpa.dupe(u8, "fatal: not a git repository");
    renderFailure(&buf.writer, gpa, "add", .{ .git_rejected = owned });
    try std.testing.expectEqualStrings("tk add: fatal: not a git repository\n", buf.written());
}

File: crates/tk/src/git/discovery.rs
//! Git `rev-parse` path discovery used to locate the Repository Store.
//!
//! `tk init` (and future commands that open the store) need Git's common
//! directory and toplevel to place or find `<git-common-dir>/tk/tk.db`.
//! This module owns the `git rev-parse` invocation and returns the discovered
//! paths as a typed result so callers can render diagnostics without this
//! module reaching for stderr itself.
//!
//! Discovery failures are a typed [`DiscoveryError`]: each variant carries its
//! own user-visible phrasing via `#[error]` (ADR-0017 stable strings), so the
//! caller renders one line without a central message catalogue.

use std::io::Write;
use std::path::{Path, PathBuf};

use thiserror::Error;

use crate::platform;
use crate::proc::{ProcError, ProcRunner};

/// Paths reported by Git for the current repository.
#[derive(Debug, Clone)]
pub struct DiscoveredPaths {
    /// Shared Git common directory; parent of the Repository Store.
    pub git_common_dir: PathBuf,
    /// Repository working-tree root; used only for display_prefix derivation.
    pub toplevel: PathBuf,
}

/// Why [`discover_paths`] could not locate the Repository Store.
///
/// Each variant's `#[error]` string is the stable user-visible phrasing
/// (ADR-0017); [`render_failure`] prints it verbatim behind the `tk <command>:`
/// prefix.
#[derive(Debug, Error)]
pub enum DiscoveryError {
    /// `git` is not on PATH (runner returned `ExecutableNotFound`).
    #[error("git not found on PATH")]
    GitMissing,
    /// Spawning `git` failed for a reason other than a missing binary.
    #[error("failed to invoke git")]
    SpawnFailed,
    /// `git rev-parse` exited non-zero. `Some` carries git's trimmed stderr,
    /// rendered verbatim; `None` (git failed silently) renders the default
    /// not-in-repo line.
    #[error("{}", .0.as_deref().unwrap_or("not in a git repository"))]
    GitRejected(Option<String>),
    /// `git rev-parse` exited zero but stdout did not contain both expected
    /// lines.
    #[error("git produced unexpected rev-parse output")]
    GitOutputUnparseable,
}

/// Run `git rev-parse --path-format=absolute --git-common-dir --show-toplevel`
/// through `runner` and classify the result.
pub fn discover_paths<R: ProcRunner + ?Sized>(
    runner: &R,
    cwd: &Path,
) -> Result<DiscoveredPaths, DiscoveryError> {
    let argv = [
        "git",
        "rev-parse",
        "--path-format=absolute",
        "--git-common-dir",
        "--show-toplevel",
    ];
    let out = match runner.run(&argv, cwd) {
        Ok(out) => out,
        Err(ProcError::ExecutableNotFound) => return Err(DiscoveryError::GitMissing),
        Err(ProcError::SpawnFailed) => return Err(DiscoveryError::SpawnFailed),
    };

    if !out.succeeded() {
        // Reuse git's stderr buffer in place — `String::from_utf8` keeps the
        // allocation when the bytes are valid UTF-8, and a non-UTF-8 stderr
        // is still useful to surface; we fall back to lossy only there.
        let stderr = String::from_utf8(out.stderr)
            .unwrap_or_else(|e| String::from_utf8_lossy(&e.into_bytes()).into_owned());
        let trimmed = stderr.trim();
        if trimmed.is_empty() {
            return Err(DiscoveryError::GitRejected(None));
        }
        return Err(DiscoveryError::GitRejected(Some(trimmed.to_string())));
    }

    // Path bytes must round-trip to a real on-disk location, so refuse rather
    // than silently lossy-decode a non-UTF-8 path. The `--path-format=absolute`
    // flag asks git to emit decoded paths; any non-UTF-8 here means the repo's
    // path uses an encoding we cannot represent as `Path` without OS-specific
    // handling (deferred to a follow-up slice).
    let Ok(stdout) = String::from_utf8(out.stdout) else {
        return Err(DiscoveryError::GitOutputUnparseable);
    };
    let mut lines = stdout.split('\n').filter(|line| !line.trim().is_empty());
    let (Some(common_raw), Some(toplevel_raw)) = (lines.next(), lines.next()) else {
        return Err(DiscoveryError::GitOutputUnparseable);
    };

    let common = normalize_native_sep(common_raw.trim().to_string());
    let toplevel = normalize_native_sep(toplevel_raw.trim().to_string());

    Ok(DiscoveredPaths {
        git_common_dir: PathBuf::from(common),
        toplevel: PathBuf::from(toplevel),
    })
}

/// Git emits forward slashes on every OS. Downstream `std::path` joins use the
/// native separator; normalise here at the discovery boundary so printed paths
/// don't come out mixed (e.g. `D:/repo/.git\tk\tk.db`).
fn normalize_native_sep(s: String) -> String {
    if platform::IS_WINDOWS {
        s.replace('/', "\\")
    } else {
        s
    }
}

/// Render a [`DiscoveryError`] to `stderr` behind a `tk <command>: ` prefix.
///
/// The error's `Display` is the stable message (ADR-0017), so this is a single
/// formatted line shared by every command that opens a Repository Store.
///
/// `command` is the bare subcommand name (`"init"`, `"add"`), without the
/// `tk ` or the trailing colon — those are formatted in.
pub fn render_failure<W: Write + ?Sized>(stderr: &mut W, command: &str, err: &DiscoveryError) {
    let _ = writeln!(stderr, "tk {command}: {err}");
}

// ---- Tests --------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use crate::proc::{ErrorInjectingRunner, FakeRunner, ProcError, RunOutput};
    use std::path::PathBuf;

    fn cwd() -> PathBuf {
        std::env::current_dir().unwrap()
    }

    #[test]
    fn returns_ok_with_both_paths_on_success() {
        let fake = FakeRunner::new();
        fake.expect(
            &["git", "rev-parse"],
            RunOutput {
                exit_code: 0,
                stdout: b"/repo/.git\n/repo\n".to_vec(),
                stderr: Vec::new(),
            },
        );
        let outcome = discover_paths(&fake, &cwd());
        match outcome {
            Ok(paths) => {
                if platform::IS_WINDOWS {
                    assert_eq!(paths.git_common_dir, PathBuf::from("\\repo\\.git"));
                    assert_eq!(paths.toplevel, PathBuf::from("\\repo"));
                } else {
                    assert_eq!(paths.git_common_dir, PathBuf::from("/repo/.git"));
                    assert_eq!(paths.toplevel, PathBuf::from("/repo"));
                }
            }
            Err(other) => panic!("expected Ok, got {other:?}"),
        }
    }

    #[test]
    fn returns_git_rejected_with_trimmed_stderr_on_exit_128() {
        let fake = FakeRunner::new();
        fake.expect(
            &["git", "rev-parse"],
            RunOutput {
                exit_code: 128,
                stdout: Vec::new(),
                stderr: b"fatal: not a git repository (or any of the parent directories): .git\n"
                    .to_vec(),
            },
        );
        match discover_paths(&fake, &cwd()) {
            Err(DiscoveryError::GitRejected(Some(msg))) => assert_eq!(
                msg,
                "fatal: not a git repository (or any of the parent directories): .git"
            ),
            other => panic!("expected GitRejected(Some), got {other:?}"),
        }
    }

    #[test]
    fn returns_git_rejected_none_when_stderr_is_whitespace() {
        let fake = FakeRunner::new();
        fake.expect(
            &["git", "rev-parse"],
            RunOutput {
                exit_code: 128,
                stdout: Vec::new(),
                stderr: b"  \r\n  ".to_vec(),
            },
        );
        assert!(matches!(
            discover_paths(&fake, &cwd()),
            Err(DiscoveryError::GitRejected(None))
        ));
    }

    #[test]
    fn returns_unparseable_when_stdout_has_no_lines() {
        let fake = FakeRunner::new();
        fake.expect(
            &["git", "rev-parse"],
            RunOutput {
                exit_code: 0,
                stdout: Vec::new(),
                stderr: Vec::new(),
            },
        );
        assert!(matches!(
            discover_paths(&fake, &cwd()),
            Err(DiscoveryError::GitOutputUnparseable)
        ));
    }

    #[test]
    fn returns_unparseable_when_stdout_has_only_one_line() {
        let fake = FakeRunner::new();
        fake.expect(
            &["git", "rev-parse"],
            RunOutput {
                exit_code: 0,
                stdout: b"/repo/.git\n".to_vec(),
                stderr: Vec::new(),
            },
        );
        assert!(matches!(
            discover_paths(&fake, &cwd()),
            Err(DiscoveryError::GitOutputUnparseable)
        ));
    }

    #[test]
    fn maps_executable_not_found_to_git_missing() {
        let runner = ErrorInjectingRunner {
            err: ProcError::ExecutableNotFound,
        };
        assert!(matches!(
            discover_paths(&runner, &cwd()),
            Err(DiscoveryError::GitMissing)
        ));
    }

    #[test]
    fn maps_spawn_failed_to_spawn_failed() {
        let runner = ErrorInjectingRunner {
            err: ProcError::SpawnFailed,
        };
        assert!(matches!(
            discover_paths(&runner, &cwd()),
            Err(DiscoveryError::SpawnFailed)
        ));
    }

    #[test]
    fn render_failure_formats_each_arm() {
        let mut buf = Vec::new();
        render_failure(&mut buf, "init", &DiscoveryError::GitMissing);
        assert_eq!(
            std::str::from_utf8(&buf).unwrap(),
            "tk init: git not found on PATH\n"
        );

        let mut buf = Vec::new();
        render_failure(&mut buf, "add", &DiscoveryError::SpawnFailed);
        assert_eq!(
            std::str::from_utf8(&buf).unwrap(),
            "tk add: failed to invoke git\n"
        );

        let mut buf = Vec::new();
        render_failure(&mut buf, "init", &DiscoveryError::GitOutputUnparseable);
        assert_eq!(
            std::str::from_utf8(&buf).unwrap(),
            "tk init: git produced unexpected rev-parse output\n"
        );

        let mut buf = Vec::new();
        render_failure(&mut buf, "init", &DiscoveryError::GitRejected(None));
        assert_eq!(
            std::str::from_utf8(&buf).unwrap(),
            "tk init: not in a git repository\n"
        );

        let mut buf = Vec::new();
        render_failure(
            &mut buf,
            "add",
            &DiscoveryError::GitRejected(Some("fatal: not a git repository".to_string())),
        );
        assert_eq!(
            std::str::from_utf8(&buf).unwrap(),
            "tk add: fatal: not a git repository\n"
        );
    }
}
